### DISCUSSION TITLE
Events the plugin receviced are same when I choose a new keys page or set a key

### DISCUSSION THREAD

**Discussion by Jonh-pc** [elgatosf/elgato#14] [General]
Hi！
While I set a new key,the events I receviced are  willappear,titleparametersdidchange,sendToplugin,sendToplugin.But I switch the other keys page（I had set a same key in this page） ,the events I recevied are similarly.I want to do some different functional development  between set a new a key  and switch a new keys page ,how could I do again?

**Comment by tiptronic** [comment]
Hi,
the keys receive a different `context`, to identify the instance's action... So - if you have 2 keys of the same action in different places, they will get the same events, but the `context` is different. That's how you can differentiate the keys and actions...

See here:
https://developer.elgato.com/documentation/stream-deck/sdk/events-received/

If you use our library, it is fairly easy to differentiate between the keys:

myAction.onKeyUp(({ action, context, device, event, payload }) => {
       // the `context` is different from key to key
	console.log('Your key code goes here!', context);
       
});

https://github.com/elgatosf/streamdeck-plugin-template/blob/main/src/com.elgato.template.sdPlugin/app.js

**Reply by Jonh-pc** [reply]
Thank you for your reply!
Another problem...... Everytime when I open the other keys page,the keys on the new page would be reload .Because all the keys's initialize functions I identify  are in the  function  named myAction,onWillappear .Could I forbid it？I want to initialize all the keys include page 1 and page 2 when I open the streamdeck.

**Comment by tiptronic** [comment]
No - the keys are loaded when they are visible.
A key basically works as an outlet, so there's no reason to have all keys loaded at all times, because you won't see them anyway.

If you need some code running while the keys are not shown, you can do that in your plugin - which runs from Stream Deck's launch until it quits.

**Reply by Jonh-pc** [reply]
Now I have some  feature to develop.I want to develop a key to switch lights's state(on-off).When I set a new key,I want to switch  lights'  state of on automatically.But when I choose page 2,the on-off keys'  state on this page get from the lights' actual state.The code I have wrote in  the function of onWillappear , and I couldn't  differentiate  between  setting-new-key‘s willappear and switch-page's willappear.

It seems that I couldn't complete this feature.

**Comment by tiptronic** [comment]
The `onWillAppear` event is always sent when the key is made available on StreamDeck (no matter if you change page or re-launch the application).

If you want the key to do something when it is added for the first time, you can check, if your plugin has 'seen' the key already. For that you could save something in a key's settings:

```
<your action>.onWillAppear(({ context, payload }) => {
  let settings = payload?.settings || {};
  if(settings.hasOwnProperty('wasAlreadyAdded') === false) { // this is a new key, because the property is missing
       settings.wasAlreadyAdded = true;  // add something so you know later, this key was added already earlier
       $SD.setSettings(context, settings);  // save settings, so next time (e.g. after a page-swipe) you can see the key was already added earlier
    } else {
      // this is a 'known' key
      // after a page-swipe the property `wasAlreadyAdded` is in your settings
     // ...
      
    }
});
```

I hope that helps

**Reply by ChekTek** [reply]
@Jonh-pc you can use this method of tracking the keys for now. We have a similar request for a `deleted` flag to determine whether a key has left the page, or if the user completely deleted the key. I will add your scenario as a feature request.

**Comment by Jonh-pc** [comment]
@tiptronic @ChekTek  It's helpful.Thanks !
